### ISSUE TITLE
Form of arguments return/values

### ISSUE THREAD

**Issue by zenhack** [ocapn/ocapn#16]
The topic of separate methods vs. a single lambda has come up before. Goblins takes the latter approach, but from what I can see still imposes some structure on the form of its argument, in particular it has a list of positional arguments and a set of keyword arguments.

Something I floated on IRC the other day was the possibility of relaxing this even further, so that from a protocol perspective, arguments & return values are each single values (in whatever our data model ends up being). In the same way that a lambda-only approach can emulate methods by looking at the first argument, a single-value approach can emulate multiple arguments by taking a list or map or the like.

I kindof like this conceptually.

We will still want to come to some agreement on a convention for the form of multi-argument functions/methods, but it seems a bit cleaner to have that at a higher level, as @cwebber has proposed for distinguishing methods, rather than baked in to the lowest levels of the protocol.

An alternative approach would be to go in the opposite direction, and mandate actually having a method name. As a practical matter, You basically always want some kind of discriminator, as otherwise you can't easily add new message types. So while conceptually elegant, the lambda-the-ultimate model may arguably be a footgun from the perspective of API stability.

**Comment by kriskowal** [comment]
I appreciate that this proposal is consistent with the wisdom of Protobuf, that not every supporting language even supports positional arguments.

**Comment by davexunit** [comment]
I had been thinking that one argument, one return (or request, response, however you want to think about it) would be the most general solution to our problem. I hadn't realized that Ian already proposed the idea over 2 years ago. This is just me speaking, not Spritely, but I think it's an idea we should examine!

**Comment by kriskowal** [comment]
I’m also not speaking on behalf of Agoric on this one. We are already quite deeply invested in variadic positional arguments, but agree this is a position worth examining.

**Comment by davexunit** [comment]
I guess I'm not seeing how this option would preclude variadic positional arguments (we use them, too) so we should talk more sometime. :)

**Comment by kriskowal** [comment]
At the August 13 OCapN meeting, a motion carried to close this issue and open a general issue to arrive at a calling convention, hinged on whether OCapN should support separate the deliver operator into forms for invoke and apply. We resolved that all method names would be either strings or symbols but not the union of either. We did not settle on whether they would be specifically strings or symbols. We resolved not to trouble ourselves over the ergonomics of the calling convention in the idioms of Objective C. Whether to embrace single-form or dual-form delivery appears to hinge on the tension between the idioms of Scheme and Haskell not mapping cleanly to separation vs OCapN symbols not mapping cleanly to JavaScript registered symbols.
